lua/gitlinker.lua

```diff
@@ -31,19 +30,20 @@ local Defaults = {
     },
   },
 
-  -- different web sites use different urls, so we want to auto bind these routers
-  --
-  -- **note**:
-  -- auto bindings only work when `router=nil` in `link` API.
-  --
-  -- github.com: `/blob`
-  -- gitlab.com: `/blob`
-  -- bitbucket.org: `/src`
+  -- router bindings
   --
+  --- @type table<"browse"|"blame", table<string, gitlinker.Router>>
   router_binding = {
```

Comment:
I would suggest a provider mapping on top router binding. Ideally one host pattern should match only one provider at a time. So user should define only one mapping which should relate the host pattern with a git provider. 

Pardon my naming convention, not sure what is the best practice for lua.
```
-- An enum somewhere on top layer
local GitProviderEnum = {
    GITHUB = "github",
    BITBUCKET = "bitbucket",
    -- Add more providers if needed
}

-- This map will help us determine the git provider from host pattern.
local git_provider_binding = {
    ["^github%.com"] = GitProviderEnum.GITHUB,
    ["^git.xyz%.com"] = GitProviderEnum.GITHUB,
    ["23.45.12.37"] = GitProviderEnum.GITHUB,
    ["^bitbucket%.org"] = GitProviderEnum.BITBUCKET,
    ["^git.abc%.com"] = GitProviderEnum.BITBUCKET,
    ["27.43.18.39"] = GitProviderEnum.BITBUCKET,
    -- Add more providers and patterns as needed
}

-- browse and blame router should be picked based on provider
router_binding = {
  [GitProviderEnum.GITHUB] = {
    browse = require("gitlinker.routers").github_browse, blame = require("gitlinker.routers").github_blame
  }
  [GitProviderEnum.BITBUCKET] = {
    browse = require("gitlinker.routers").bitbucket_browse, blame = require("gitlinker.routers").bitbucket_blame
  }
}
```

Reply:
Internally we can have default binding like below which can be extended by user.
```
local default_git_provider_binding = {
    ["^github%.com"] = GitProviderEnum.GITHUB,
    ["^bitbucket%.org"] = GitProviderEnum.BITBUCKET,
    -- Default bindings
}
```

Reply:
I meant we should expose only `git_provider_binding` (host url to git service provider mapping) to user because usually only url changes when on campus setup is opted for git provider.

We can keep `router_binding` abstract and should avoid exposing to end user, browse and blame router should be internal and should be implemented when new provider support is added to the plugin.

Reply:
hi @G0V1NDS ,

1. if only expose `git_provider_binding` to user, there're requirements from self-host users, that try to customize the url generation. only `enum` seems cannot achieve this requirement. I also want to try to cover these use cases. please see: https://github.com/linrongbin16/gitlinker.nvim/issues/114.
2. hide the `browse` and `blame` API as internal - agreed, I have also did something similar. In latest changes, I only expose the `router` option (I renamed `router_binding` to `router`) to bind regex-based (lua pattern) host to a specific router implementation. so users won't touch the internal `browse` and `blame` API.

Here's a summary about what I did yesterday and today:
- Provide the `GitLink` command, drop off default key mappings (it's another big break change, so versioning is up to v4.0.0 now). I design this command as:
  - `GitLink (browse)` to copy link to clipboard.
  - `GitLink! (browse)` to open link in browser.
  - `GitLink blame` to copy `/blame` url to clipboard.
  - `GitLink! blame` to open `/blame` url to browser.

  so in currently workflow, I found that there's no more need to expose these lua APIs `link`, `routers`, `actions` to user anymore. if contains bang `!` defines the action, and an optional argument `browse` (default) and `blame` defines the router type.

- Rewrite the `router_binding` option with:
  - Rename it to `router`, this is just for simplicity.
  - Allow the hash map value can be a string template, which makes user easier to write a customize url. While functions such as `github_browse`/`github_blame` are still allowed.


Please see:
- [README - usage](https://github.com/linrongbin16/gitlinker.nvim#usage)
- provide `GitLink` command: https://github.com/linrongbin16/gitlinker.nvim/pull/120
- drop off default key mappings!: https://github.com/linrongbin16/gitlinker.nvim/pull/126
- hide `browse`/`blame` as internal API: https://github.com/linrongbin16/gitlinker.nvim/pull/127
- support url string template: https://github.com/linrongbin16/gitlinker.nvim/pull/128

Reply:
I lost the track, 2 major versions in same day 😮

I was about to make a separate issue for keybind, but now it won't be needed. 
With previous implementation, I wasn't able to lazy load the plugin because I had to load the plugin first to import `require("gitlinker.actions").system`. 

Now with `GitLink` cmd I can easily lazy load the plugin:
```lua
{
  "linrongbin16/gitlinker.nvim",
  cmd = { "GitLink" },
  config = true,
}
```
Appreciate your for your dedication. ❤️

Reply:
After add the `GitLink`, I found the whole workflow becomes quite easy, not more lua API expose to users, unless they want to do heavy customization.